lib/evaluate/shape.cc

```diff
@@ -0,0 +1,357 @@
+  if (symbol != nullptr) {
+    return GetShape(*symbol);
+  } else {
+    return std::nullopt;
+  }
+}
+
+std::optional<Shape> GetShape(const BaseObject &object) {
+  if (const Symbol * symbol{object.symbol()}) {
+    return GetShape(*symbol);
+  } else {
+    return Shape{};
+  }
+}
+
+std::optional<Shape> GetShape(const Component &component) {
+  const Symbol &symbol{component.GetLastSymbol()};
+  if (symbol.Rank() > 0) {
+    return GetShape(symbol, &component);
+  } else {
+    return GetShape(component.base());
+  }
+}
+
+std::optional<Shape> GetShape(const ArrayRef &arrayRef) {
+  Shape shape;
+  const Symbol &symbol{arrayRef.GetLastSymbol()};
+  const Component *component{std::get_if<Component>(&arrayRef.base())};
+  int dimension{0};
+  for (const Subscript &ss : arrayRef.subscript()) {
+    if (ss.Rank() > 0) {
+      shape.emplace_back(GetExtent(ss, symbol, component, dimension));
+    }
+    ++dimension;
+  }
+  if (shape.empty()) {
+    return GetShape(arrayRef.base());
+  } else {
+    return shape;
+  }
+}
+
+std::optional<Shape> GetShape(const CoarrayRef &coarrayRef) {
+  Shape shape;
+  SymbolOrComponent base{coarrayRef.GetBaseSymbolOrComponent()};
+  const Symbol &symbol{coarrayRef.GetLastSymbol()};
+  const Component *component{std::get_if<Component>(&base)};
+  int dimension{0};
+  for (const Subscript &ss : coarrayRef.subscript()) {
+    if (ss.Rank() > 0) {
+      shape.emplace_back(GetExtent(ss, symbol, component, dimension));
+    }
+    ++dimension;
+  }
+  if (shape.empty()) {
+    return GetShape(coarrayRef.GetLastSymbol());
+  } else {
+    return shape;
+  }
+}
+
+std::optional<Shape> GetShape(const DataRef &dataRef) {
+  return GetShape(dataRef.u);
+}
+
+std::optional<Shape> GetShape(const Substring &substring) {
+  if (const auto *dataRef{substring.GetParentIf<DataRef>()}) {
+    return GetShape(*dataRef);
+  } else {
+    return std::nullopt;
+  }
+}
+
+std::optional<Shape> GetShape(const ComplexPart &part) {
+  return GetShape(part.complex());
+}
+
+std::optional<Shape> GetShape(const ActualArgument &arg) {
+  return GetShape(arg.value());
+}
+
+std::optional<Shape> GetShape(const ProcedureRef &call) {
+  if (call.Rank() == 0) {
+    return Shape{};
+  } else if (call.IsElemental()) {
+    for (const auto &arg : call.arguments()) {
+      if (arg.has_value() && arg->Rank() > 0) {
+        return GetShape(*arg);
+      }
+    }
+  } else if (const Symbol * symbol{call.proc().GetSymbol()}) {
+    return GetShape(*symbol);
+  } else if (const auto *intrinsic{
+                 std::get_if<SpecificIntrinsic>(&call.proc().u)}) {
+    if (intrinsic->name == "shape" || intrinsic->name == "lbound" ||
+        intrinsic->name == "ubound") {
+      return Shape{MaybeExtent{
+          ExtentExpr{call.arguments().front().value().value().Rank()}}};
+    } else if (intrinsic->name == "reshape") {
+      if (call.arguments().size() >= 2 && call.arguments().at(1).has_value()) {
+      }
```

Comment:
Is this a TODO?

Reply:
No, it's something that I intended to get to before review, sorry; will complete it now.